osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs

```diff
@@ -42,330 +31,33 @@ public void ApplyToBeatmap(IBeatmap beatmap)
             if (!(beatmap is OsuBeatmap osuBeatmap))
                 return;
 
-            var hitObjects = osuBeatmap.HitObjects;
-
             Seed.Value ??= RNG.Next();
 
             rng = new Random((int)Seed.Value);
 
-            var randomObjects = randomiseObjects(hitObjects);
-
-            applyRandomisation(hitObjects, randomObjects);
-        }
+            var positionModifier = new OsuHitObjectPositionModifier(osuBeatmap.HitObjects);
```

Comment:
I'm not super sure how to feel about the API of this. In short:

* First, a "position modifier" is constructed from the list of beatmaps, which does all of the generation work. It happens to inadvertently borrow the hitobjects from the beatmap and store them to its internal list.
* Then to introduce changes, the internal-but-public-ish state of this "position modifier" is changed to apply the changes random wishes to apply to map objects.
* Finally this same modifier instance, whose internal-but-public-ish state was just modified, is used to mutate hitobjects on the internal list, which ends up modifying the beatmap objects by extensions.

There's a significant number of invasive or not-clearly-marked operations in this logic as proposed, via silent assumptions and reference borrowing. Is there a reason why this application process couldn't be computationally pure? As in, have a static, stateless method that takes in a list of hitobjects and returns a list of `IObjectPositionInfo`s back, and a second static method that takes a list of unmodified hitobjects and a list of `IObjectPositionInfo`s (that could have been messed with to the implementor's wishes), and applies the object position adjustments onto the objects?

Reply:
I've converted the class into a bunch of static methods and put them in `OsuHitObjectGenerationUtils` (in a separate file because it's too big). To be honest, I'm not sure if static methods look cleaner, but a definite upside of using them is that implementors can now add/remove items from the list of `ObjectPositionInfo`s before handing it to `RepositionHitObjects(...)`. I've changed the signature of `ApplyModifications` (now called `RepositionHitObjects`) to account for this.